solidity/contracts/Chainlinked.sol

```diff
@@ -24,31 +24,21 @@ contract Chainlinked {
     string _callbackFunctionSignature
   ) internal pure returns (ChainlinkLib.Run memory) {
     ChainlinkLib.Run memory run;
-    Buffer.init(run.buf, 128);
-    run.jobId = _jobId;
-    run.callbackAddress = _callbackAddress;
-    run.callbackFunctionId = bytes4(keccak256(_callbackFunctionSignature));
-    run.buf.startMap();
-
-    return run;
+    return run.initialize(_jobId, _callbackAddress, _callbackFunctionSignature);
   }
 
   function chainlinkRequest(ChainlinkLib.Run memory _run, uint256 _wei)
     internal
     returns(bytes32)
   {
-    bytes32 requestId = keccak256(this, requests++);
-    bytes memory requestDataABI = abi.encodeWithSelector(
-      oracleFid,
-      clArgsVersion,
-      _run.jobId,
-      _run.callbackAddress,
-      _run.callbackFunctionId,
-      requestId,
-      _run.close());
-    require(link.transferAndCall(oracle, _wei, requestDataABI));
-
-    return requestId;
+    _run.externalId = keccak256(this, requests++);
+    _run.close();
+    require(link.transferAndCall(
+      oracle,
+      _wei,
+      _run.encodeForOracle(oracleFid, clArgsVersion)));
```

Comment:
At first glance, it seems strange to have to pass the `oracleFid` and `clArgsVersion` into `encodeForOracle`. However, I understand that you want consumers of different versions using the same ChainlinkLib. The ironic thing though is that if the `oracleFid` changed, you would still need to change `encodeForOracle`:

```solidity
return abi.encodeWithSelector(
      _oracleFid,
      _clArgsVersion,
      self.jobId,
      self.callbackAddress,
      self.callbackFunctionId,
      self.externalId,
      self.buf.buf);
```

If the types or orders change, that encoding line still has to change.

Reply:
```solidity
  function requestData(
    uint256 _version,
    bytes32 _jobId,
    address _callbackAddress,
    bytes4 _callbackFunctionId,
    bytes32 _externalId,
    bytes _data
  )
```

If any types change, we would need a recompilation of the contract and the oracleFid would change (different keccak(`STRINGCHANGE`)), therefore both the `oracleFid` and the line quoted above would probably change. If, somehow, we change the type but keep the name, it would still need a recompilation and deployment. I do think it's better in Chainlinked. Let's chat offline.